tunnel/intra/doh.go

```diff
@@ -0,0 +1,323 @@
+				server, _, _ = net.SplitHostPort(addr.String())
+			}
+		},
+	}
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), &trace))
+
+	const mimetype = "application/dns-message"
+	req.Header.Set("Content-Type", mimetype)
+	req.Header.Set("Accept", mimetype)
+	req.Header.Set("User-Agent", "Intra")
+	httpResponse, err := t.client.Do(req)
+	if err != nil {
+		qerr = &queryError{SendFailed, err}
+		return
+	}
+	if httpResponse.StatusCode != http.StatusOK {
+		err := fmt.Errorf("HTTP request failed: %d", httpResponse.StatusCode)
+		qerr = &queryError{HTTPError, err}
+		return
+	}
+	response, err = ioutil.ReadAll(httpResponse.Body)
+	httpResponse.Body.Close()
+	if err != nil {
+		qerr = &queryError{BadResponse, err}
+		return
+	}
+	// Restore the query ID.
+	q[0], q[1] = id0, id1
+	if len(response) >= 2 {
+		response[0], response[1] = id0, id1
+	} else {
+		qerr = &queryError{BadResponse, fmt.Errorf("Response length is %d", len(response))}
+		return
+	}
+	// Record a working IP address for this server
+	t.ips.Get(httpResponse.Request.URL.Hostname()).Confirm(server)
+	return
+}
+
+func (t *transport) Query(q []byte) ([]byte, error) {
+	before := time.Now()
+	response, server, err := t.doQuery(q)
+	after := time.Now()
+	if t.listener != nil {
+		latency := after.Sub(before)
+		status := Complete
+		var qerr *queryError
+		if errors.As(err, &qerr) {
+			status = qerr.status
+		}
+		t.listener.OnDNSTransaction(&DNSSummary{
+			Latency:  latency.Seconds(),
+			Query:    q,
+			Response: response,
+			Server:   server,
+			Status:   status,
+		})
+	}
+	return response, err
+}
+
+func (t *transport) GetURL() string {
+	return t.url
+}
+
+// Perform a query using the transport, and send the response to the writer.
+func forwardQuery(t DNSTransport, q []byte, c io.Writer) error {
+	resp, err := t.Query(q)
+	if err != nil {
+		return err
+	}
+	rlen := len(resp)
+	if rlen > math.MaxUint16 {
+		return fmt.Errorf("Oversize response: %d", rlen)
+	}
+	// Use a combined write to ensure atomicity.  Otherwise, writes from two
+	// responses could be interleaved.
+	rlbuf := make([]byte, rlen+2)
+	binary.BigEndian.PutUint16(rlbuf, uint16(rlen))
+	copy(rlbuf[2:], resp)
+	n, err := c.Write(rlbuf)
+	if err != nil {
+		return err
+	}
+	if int(n) != len(rlbuf) {
+		return fmt.Errorf("Incomplete response write: %d < %d", n, len(rlbuf))
+	}
+	return nil
+}
+
+// Perform a query using the transport, send the response to the writer,
+// and close the writer if there was an error.
+func forwardQueryAndCheck(t DNSTransport, q []byte, c io.WriteCloser) {
+	if err := forwardQuery(t, q, c); err != nil {
+		fmt.Printf("Query forwarding failed: %v\n", err)
+		c.Close()
+	}
+}
+
+// Accept a DNS-over-TCP socket from a stub resolver, and connect the socket
+// to this DNSTransport.
+func Accept(t DNSTransport, c io.ReadWriteCloser) {
+	qlbuf := make([]byte, 2)
+	for n, err := c.Read(qlbuf); err == nil && n == 2; n, err = c.Read(qlbuf) {
```

Comment:
Do we want to log query lenght read failures? Can we, using this construct?

Reply:
OK, done.